### ISSUE TITLE
Interact's abbreviation can only handle basic objects

### ISSUE THREAD

**Issue by mocquin** [jupyter-widgets/ipywidgets#3015]
Hello there, I looked if it would be possible to make a custom class handled by `ipywidgets.interact`, the same way `Int`, `Float`, `Tuple` and friends are natively handled.

# Lay the foundation
Looking at the code, I understand the following : 
 - the objects passed to `interact`, if they are not widgets instances already, are looked up to be converted to widgets using "abbreviations" with `widget_from_abbrev` :
https://github.com/jupyter-widgets/ipywidgets/blob/fcadca8fbdc18ef1d8dc5694e8e6e759af6c7749/ipywidgets/widgets/interaction.py#L310-L311
 - say a single object is passed to interact, this will rely on `widget_from_single_value` : 
https://github.com/jupyter-widgets/ipywidgets/blob/fcadca8fbdc18ef1d8dc5694e8e6e759af6c7749/ipywidgets/widgets/interaction.py#L346
 - here, we can see that few data types are handled, basically strings, bools, and `numbers.Real` and `numbers.Integral`
 - if the object passed to interact is not of these types, we will get the `ValueError` :
https://github.com/jupyter-widgets/ipywidgets/blob/fcadca8fbdc18ef1d8dc5694e8e6e759af6c7749/ipywidgets/widgets/interaction.py#L300

# The issue
Now my problem is that the object I want to be handled by `interact` is not of those types (and even if I modify my class a bit to inherit from `numbers.Real`, I get exceptions raised in `_get_min_max_value` for class-related reasons). 

Is there a way to "register" a class with a function that would return the appropriate widget (that would be provided) and make the class work with `interact` seamlessly ? I'd expect something like `ipywidgets.interact.register[MyClass]= MyClassWidget`.

**Comment by ianhi** [comment]
I don't think there's currently a way to do this though it seems like a reasonable feature to have. The ability to register a factory function that would return a ValueWidget for different types.

Everything below this line may be very bad practice (but I honestly don't know) so use at your own risk.

----
Here's my suggestion for how to accomplish this. Just overwrite the ipywidgets interact function in your `__init__.py` with a custom wrapper around it. So you'd do something like:
```python
from ipywidgets import interact as orig_interact
import ipywidgets
def my_interact(_InteractFactory__interact_f=None, **kwargs):
    for k,v in kwargs.items():
        if isinstance(v, YOUR_CLASS):
            # create whatever widget you want
            some_widget = ....
            kwargs[k] = some_widget
    return orig_interact(_InteractFactory__interact_f, **kwargs)
my_interact.__doc__ = orig_interact.__doc__
ipywidgets.interact = my_interact
```

**Comment by ianhi** [comment]
I think the reason that feels so sketchy to me is that it's totally unregulated. So if another library does something similar it could easily turn into an enormous mess that may behave differently depending on import order.

**Comment by mocquin** [comment]
Agreed. Ideally, this feature would be totally on ipywidgets side, so any library that wants to use this would only have to write few lines of code. 

An even better approach than the "register" one above, would be like numpy's possibility to override `ufuncs` and `functions` : basically the object could implement a method called `__widgetify_abbrev__` thant handles the conversion to a valid widget, and the presence of this `__widgetify_abbrev__` method is tested in `ipywidgets.widget_from_single_value` for eg : if the method is present, it is called to cast the object into its widget representation (kinda like a hook). This way only a method have to be put in the class and nothing else (no registering at startup, no wrapper). Also, it would work when mixing any number of libraries (as long as they use this approach)
